### ISSUE TITLE
No music files visible in 6.2 beta2 [MySQL/MariaDB]

### ISSUE THREAD

**Issue by joshuaboniface** [Libresonic/libresonic#319]
Recently upgraded to 6.2 beta2, using embedded JDBC and the MariaDB connector under tomcat8. This issue is appearing on a fresh install (new VM and fresh install) including a fresh database, but was also happening after my upgrade from 6.2 beta1 (built a new VM to make sure it was reproducible).

Everything appears OK during configuration and scans, the database is populated and looks correct, and the full list of artists appears on the sidebar. However, I'm not seeing anything in the Artist views (so no Albums), nor in any Album views that pop up on the home page (no Songs). In short, it looks like no media files are found, even though they're present in the database. Sometimes during the scan (and in particular if the scan is interrupted) I can see certain albums, but never all of them, they won't play, and as soon as the scan completes I'm left with no visible Songs.

From my reading (mostly of old Subsonic posts) and experience here, there are two causes: permissions, and transcoding. However I've been reconfiguring both and wracking my brain all weekend to no avail.

I've attempted the library permission off/on trick multiple times with no success, all as the `admin` user, so I'm fairly convinced that it isn't a permissions thing, and if it is, it's very deep. I've tried logging in/out, restarting tomcat, and restarting the VM between on and off, all with no effect.

My library is entirely FLAC, so I've also tried three different transcoding settings: first with the default `ffmpeg` (including the link in the `/var/libresonic/transcode` directory), and second and third with a custom `flac`+`lame` config which I had used previously (both with removing all the `ffmpeg` ones, and keeping them but removing `flac` types), however no changes I make here seem to correct the problem. I had originally moved to this `flac`+`lame` config (specifically without any `ffmpeg` ones) precisely because of this error on 6.2 beta1, which it fixed, but on beta2 it looks to no longer have an effect.

It's worth mentioning I think that I do get lots of `Missing /WEB-INF/web.xml` messages, but they're not fatal. My Catalina config is a `ROOT.xml` out of the Tomcat docs to provide Libresonic on `/`, but otherwise there's nothing special there that I can tell. There are no errors to speak of in the log, and I'm not sure if there's a way to provide more debug logging. I'm willing to provide any additional details that might be needed. I do hope this is some config error on my side, but so far I've kept this as close to the docs as I can (and moreso than my config for 6.2 beta1 and earlier) and can't seem to avoid it, so I'm worried it might be more systemic.

**Comment by joshuaboniface** [comment]
As a test, I installed a new instance of Libresonic using my same procedure, but with a single album in the library. The results are the same though, so here's all the details I can muster:

Catalina config:
```
$ cat /etc/tomcat8/Catalina/localhost/ROOT.xml
<!-- Libresonic Tomcat configuration -->
<!-- Ansible managed - last modified on 2017-03-26 18:42:43 -->
<Context 
        docBase="/var/libresonic/libresonic.war"
        path=""
        reloadable="true"
/>
```

The directory listing of the one media file:

```
$ ls -alh /srv/media/Music.test/Jon\ Anderson/Olias\ of\ Sunhillow/
total 239M
drwxr-xr-x 2 tomcat8 media   4.0K Mar 20 11:56 .
drwxr-xr-x 3 tomcat8 media     39 Mar 20 00:30 ..
-rw-r--r-- 1 tomcat8 media    15M Mar 20 00:31 01 Ocean Song.flac
-rw-r--r-- 1 tomcat8 media    22M Mar 20 00:31 02 Meeting (Garden of Geda) _ Sound Out the Galleon.flac
-rw-r--r-- 1 tomcat8 media    22M Mar 20 00:31 03 Dance of Raynart _ Olias (To Build the Moorglade).flac
-rw-r--r-- 1 tomcat8 media    39M Mar 20 00:31 04 Qoquaq ën transic _ Naon _ Transic tö.flac
-rw-r--r-- 1 tomcat8 media    21M Mar 20 00:31 05 Flight of the Moorglade.flac
-rw-r--r-- 1 tomcat8 media    32M Mar 20 00:31 06 Solid Space.flac
-rw-r--r-- 1 tomcat8 media    66M Mar 20 00:31 07 Moon Ra _ Chords _ Song of Search.flac
-rw-r--r-- 1 tomcat8 media    25M Mar 20 00:31 08 To the Runner.flac
-rw-r--r-- 1 tomcat8 media    60K Mar 20 00:31 albumart.jpg
-rw-r--r-- 1 tomcat8 tomcat8 254K Mar 20 11:56 cover.png
```

Configured `libresonic.properties` (sanitized):
```
$ sudo cat /var/libresonic/libresonic.properties
# Libresonic preferences.  NOTE: This file is automatically generated. Do not modify while application is running

MediaLibraryStatistics=1 8 8 250243755 2664
LastScanned=1490631144324
DatabaseMysqlMaxlength=512
DatabaseConfigType=EMBED
DatabaseConfigEmbedDriver=org.mariadb.jdbc.Driver
DatabaseConfigEmbedPassword=<PASS>
DatabaseConfigEmbedUrl=jdbc:mysql://<URL>/libresonic_test?sessionVariables=sql_mode=ANSI_QUOTES
DatabaseConfigEmbedUsername=libresonic
DatabaseUsertableQuote=
SettingsChanged=1490631129490
IndexCreationInterval=1
IndexCreationHour=3
FastCacheEnabled=true
OrganizeByFolderStructure=true
DownsamplingCommand4=ffmpeg -i %s -map 0:0 -b:a %bk -v 0 -f mp3 -
HlsCommand3=ffmpeg -ss %o -t %d -i %s -async 1 -b:v %bk -s %wx%h -ar 44100 -ac 2 -v 0 -f mpegts -c:v libx264 -preset superfast -c:a libmp3lame -threads 0 
```

Startup log from Tomcat's `catalina.out`:

```
INFO: Deploying configuration descriptor /etc/tomcat8/Catalina/localhost/ROOT.xml
  _      _ _                               _       ______
 | |    (_) |                             (_)      \ \ \ \
 | |     _| |__  _ __ ___  ___  ___  _ __  _  ___   | | | |
 | |    | | '_ \| '__/ _ \/ __|/ _ \| '_ \| |/ __|  | | | |
 | |____| | |_) | | |  __/\__ \ (_) | | | | | (__   | | | |
 |______|_|_.__/|_|  \___||___/\___/|_| |_|_|\___|  | | | |
                                                   /_/_/_/
                                

2017-03-27 16:15:39.243  INFO --- o.l.player.boot.Application              : Starting Application on ls2.domain.net with PID 1782 (/var/lib/tomcat8/webapps/ROOT/WEB-INF/classes/org/libresonic/player/boot/Application.class started by tomcat8 in /var/lib/tomcat8)
2017-03-27 16:15:39.258  INFO --- o.l.player.boot.Application              : The following profiles are active: embed
2017-03-27 16:15:44.149  INFO --- o.l.p.service.SettingsService            : Java: 1.8.0_111, OS: Linux
2017-03-27 16:16:01.108  INFO --- liquibase                                : Successfully acquired change log lock
2017-03-27 16:16:04.038  INFO --- liquibase                                : Reading from libresonic_test.DATABASECHANGELOG
2017-03-27 16:16:04.471  INFO --- liquibase                                : Successfully released change log lock
2017-03-27 16:16:04.494  INFO --- liquibase                                : Successfully acquired change log lock
2017-03-27 16:16:04.510  INFO --- liquibase                                : Reading from libresonic_test.DATABASECHANGELOG
2017-03-27 16:16:04.714  INFO --- liquibase                                : Successfully released change log lock
2017-03-27 16:16:05.646  INFO --- o.l.player.dao.PlayerDao                 : Deleted 1 player(s) that haven't been used after Thu Jan 26 16:16:05 UTC 2017
2017-03-27 16:16:05.774  INFO --- o.l.p.s.MediaScannerService              : Automatic media library scanning scheduled to run every 1 day(s), starting at Tue Mar 28 03:00:00 UTC 2017
2017-03-27 16:16:05.900  INFO --- o.l.p.service.PodcastService             : Automatic Podcast update scheduled to run every 24 hour(s), starting at Mon Mar 27 16:21:05 UTC 2017
2017-03-27 16:16:06.064  INFO --- o.l.player.service.UPnPService           : Starting UPnP service...
2017-03-27 16:16:06.766  INFO --- o.l.player.service.UPnPService           : Disabling UPnP/DLNA media server
2017-03-27 16:16:06.766  INFO --- o.l.player.service.UPnPService           : Starting UPnP service - Done!
2017-03-27 16:16:14.724  INFO --- o.l.player.boot.Application              : Started Application in 38.633 seconds (JVM running for 65.571)
Mar 27, 2017 4:16:14 PM org.apache.catalina.startup.HostConfig deployDescriptor
INFO: Deployment of configuration descriptor /etc/tomcat8/Catalina/localhost/ROOT.xml has finished in 64,374 ms
Mar 27, 2017 4:16:14 PM org.apache.coyote.AbstractProtocol start
INFO: Starting ProtocolHandler ["http-nio-8080"]
Mar 27, 2017 4:16:14 PM org.apache.catalina.startup.Catalina start
INFO: Server startup in 64459 ms
2017-03-27 16:16:15.641  INFO --- o.l.p.f.BootstrapVerificationFilter      : Servlet container: Apache Tomcat/8.0.14 (Debian)
2017-03-27 16:16:15.688  WARN --- o.d.impl.DefaultPageNormalizer           : Missing /WEB-INF/web.xml
2017-03-27 16:16:15.844  INFO --- o.l.player.dao.PlayerDao                 : Created player 1.
2017-03-27 16:16:18.455  INFO --- o.l.p.service.VersionService             : Resolved local Libresonic version to: 6.2.beta2
```

The contents of the database table `transcoding2`:
```
mysql> select * from transcoding2;
+----+------+----------------+---------------+------------------------------------+--------------------------+-------+----------------+
| id | name | source_formats | target_format | step1                              | step2                    | step3 | default_active |
+----+------+----------------+---------------+------------------------------------+--------------------------+-------+----------------+
|  1 | flac | flac           | mp3           | flac --silent --decode --stdout %s | lame --silent -h -b %b - | NULL  |               |
+----+------+----------------+-
```

Hopefully this is enough to point in a particular direction as this seems to have no obvious cause to me.

**Comment by muff1nman** [comment]
Sorry for the short comment -- I'll take a closer look later. But at first glance I think the first thing to look at is why your media files are not marked as present.

**Comment by joshuaboniface** [comment]
Well that's interesting, I thought I noticed that during my original testing and found that they were set active (to `1` in that column) and just not showing (MariaDB display bug perhaps?), but with my current test, they were indeed set to `0` instead. Updated them and the files now show.

Is there a way to enable debugging to see why they might be set not present?

**Comment by joshuaboniface** [comment]
Well this is curious: while I'm scanning, the entries are `1`, but as soon as the scan finishes they are set to `0`. I do see the log entries about `Marking non-present files/artists/albums` but no indication of which files are set or why. This explains why I was seeing some files when a scan was interrupted.

**Comment by muff1nman** [comment]
Is this in MariaDB 10.2? I'm wondering if there is some date mishap going on there.

**Comment by joshuaboniface** [comment]
Yep, MariaDB 10.2.4 (build 10.2.4-MariaDB-10.2.4+maria~jessie).

**Comment by muff1nman** [comment]
@joshuaboniface See if you can give the latest on develop a shot.

I was able to reproduce the present being set to false after a scan, however for some reason my songs were still getting displayed. Regardless, I think the change I made should fix part of the problem.

**Comment by joshuaboniface** [comment]
Sure thing I'll give it a shot and report.

**Comment by joshuaboniface** [comment]
That does look to have fixed it, thanks @muff1nman!

**Comment by joshuaboniface** [comment]
So that fixed the library problem, but at this point I'm getting a whole different issue (though it might be related and hence why I'm keeping it here).

My config is identical to that shown above still.

However, playing music in the web player does not seem to work. Mobile player (using the official Subsonic app) works fine. However in web, the items are added to the play queue and then stop. No transcoder process is spawned at all. This seems to be affecting every file.

I've tweaked with the transcoder settings a bit more (going back between the default `ffmpeg` transcoder and my `flac`+`lame` encoder, but that doesn't seem to fix it, and mobile works with both.

Seems too much of a coincidence to not be related to the media scanning trouble, any thoughts?

**Comment by muff1nman** [comment]
You can add `-Ddebug=true` for more output although it still may not be sufficient for finding the problem. We may need to add some more logging.

My hunch is it's another MariaDB edge case which I'll have to work with you find.

**Comment by joshuaboniface** [comment]
I'm thinking the same, I'll see if debug gives up any more information while I try to replicate this in a test environment.

**Comment by muff1nman** [comment]
Any activity in the logs? How about when you open the developer tools on the browser and monitor the network activity? Using MariaDB so far and I havent run into any issues yet, but my test library doesn't contain any flac songs.

**Comment by joshuaboniface** [comment]
Well, it looks like this is on me and my crazy config. I did see the crashes mentioned in #323 there though, but as a side effect of my testing, so I didn't keep the log (assumed it was me). If I see those again I'll post there.

Part of the problem was a messed up transcoder setting, which I caused during debugging of the second part, which is that I'm loading the player as insecure HTTP and Firefox 52 is refusing it - this seems to be the crux of my "doesn't play" issue, but my transcoding error obscured that. Should have checked the developer console sooner! If I launch the insecure version directly it plays fine, and the secure version plays fine in Chrome (which doesn't do the blocking FF does).

This is actually totally unexpected, and seems far enough removed from the original issue here than I'm going to make another addressing it specifically. I think this issue as described is solved, and the last part a red herring. Thanks again @muff1nman!
